encryptor/queryDataEncryptor.go

```diff
@@ -0,0 +1,263 @@
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package encryptor
+
+import (
+	"github.com/cossacklabs/acra/decryptor/base"
+	"github.com/cossacklabs/acra/sqlparser"
+	"github.com/sirupsen/logrus"
+	"reflect"
+)
+
+// QueryDataEncryptor parse query and encrypt raw data according to TableSchemaStore
+type QueryDataEncryptor struct {
+	schemaStore TableSchemaStore
+	encryptor   DataEncryptor
+	clientID    []byte
+	dataCoder   DBDataCoder
+}
+
+// NewMysqlQueryEncryptor create QueryDataEncryptor with MySQLDBDataCoder
+func NewMysqlQueryEncryptor(schema TableSchemaStore, clientID []byte, dataEncryptor DataEncryptor) (*QueryDataEncryptor, error) {
+	return &QueryDataEncryptor{schemaStore: schema, clientID: clientID, encryptor: dataEncryptor, dataCoder: &MysqlDBDataCoder{}}, nil
+}
+
+// NewPostgresqlQueryEncryptor create QueryDataEncryptor with PostgresqlDBDataCoder
+func NewPostgresqlQueryEncryptor(schema TableSchemaStore, clientID []byte, dataEncryptor DataEncryptor) (*QueryDataEncryptor, error) {
+	return &QueryDataEncryptor{schemaStore: schema, clientID: clientID, encryptor: dataEncryptor, dataCoder: &PostgresqlDBDataCoder{}}, nil
+}
+
+// encryptInsertQuery encrypt data in insert query in VALUES and ON DUPLICATE KEY UPDATE statements
+func (encryptor *QueryDataEncryptor) encryptInsertQuery(insert *sqlparser.Insert) (bool, error) {
+	tableName := sqlparser.String(insert.Table.Name)
+	schema := encryptor.schemaStore.GetTableSchema(tableName)
+	if schema == nil {
+		// unsupported table, we have not schema and query hasn't columns description
+		logrus.Debugf("Hasn't schema for table %s", tableName)
+		return false, nil
+	}
+
+	var columnsName []string
+	if len(insert.Columns) > 0 {
+		columnsName = make([]string, 0, len(insert.Columns))
+		for _, col := range insert.Columns {
+			columnsName = append(columnsName, sqlparser.String(col))
+		}
+	} else if len(schema.Columns) > 0 {
+		columnsName = schema.Columns
+	}
+
+	changed := false
+
+	if len(columnsName) > 0 {
+		switch rows := insert.Rows.(type) {
+		case sqlparser.Values:
+			for _, valTuple := range rows {
+				// collect values per column
+				for j, value := range valTuple {
+					columnName := columnsName[j]
+					if changedValue, err := encryptor.encryptExpression(value, schema, columnName); err != nil {
+						logrus.WithError(err).Errorln("Can't encrypt expression")
+						return changed, err
+					} else if changedValue {
+						changed = true
+					}
+				}
+			}
+		}
+	}
+
+	if len(insert.OnDup) > 0 {
```

Comment:
onDup? duplicate key?

Reply:
what to do if database recognize duplication. it's like update on duplicate or upsert.